### ISSUE TITLE
About the training of flow matching

### ISSUE THREAD

**Issue by huskyachao** [FunAudioLLM/CosyVoice#308]
Hey, guys. I am trying to train the flow model from scratch recently. 

But I am a bit confused about the training pipeline of the flow model. **As suggested by , the flow can be trained just change the param `model` in the run.sh script to 'flow', which will initalize the whole flow model**, e.g., `flow: !new:cosyvoice.flow.flow.MaskedDiffWithXvec`. **I am confused that the loss function of the flow only take the loss of `cosyvoice.flow.flow.MaskedDiffWithXvec.decoder ` into consideration** without any from other modules of the flow (e.g., encoder, length_regulator). 

**_Question:_**
1. Is this manner suitable for the flow training?  
2. How did parameters of these modules (except the decoder) update if we just initialize the whole flow model? 
3. Should we freeze partial of the flow (e.g., the encoder, length_regulator, and nn.Embedding()) to train the flow model?

**Comment by aluminumbox** [comment]
do not freeze encoder, reduce lr or increase batch size

**Comment by JohnHerry** [comment]
I compared the flow model config of cosyvoice.yaml and cosyvoice.fromscratch.yaml;   the former is used to fine-tune on the pretrained model, and the latter is used to train from scratch on small dataset.  I found that the small model (trained on small dataset) config compared to the big model (the open source one configed as the cosyvoice.yaml),  the parameter reduction mainly occurs on the Conformer Encoder part,  decrease from 6 block to 3 bloks,  but on the decoder part, the ConditionalDecoder [worked as ODE estimator ] config,  it only reduce 4 blocks of mid_blocks;  Why there keeps heavy paremeter config on the ODE estimator?  If I want to further reduce the model parameters, where can I cut more?

**Comment by JohnHerry** [comment]
关于flow模型的 length_regulator, 没有太明白一些细节。 理解上是由于speech_token抽取的时候，S3以50的码率编码，一个码代表了320个样点； mel谱是用256的步长编码，一个梅尔帧256个样点，16K数据的帧率就是62.5； 现在flow就是要将50码率的东西，转为62.5码率的东西； length_regulator 就是对50的序列内部插值，插值到62.5的码率下长度。感觉length_regulator 里一堆卷积变换的，就是对硬插值后得到的结果做一下平滑。但是length_regulator的inference里，又是分频带又是组合的，这段逻辑在做什么啊？

**Comment by JohnHerry** [comment]
正常的情况下 flow模型训练一步需要多少时间啊？需要训练多少步？ 我们这里训练怎么这么慢呢？

**Comment by JohnHerry** [comment]
> Hey, guys. I am trying to train the flow model from scratch recently.
> 
> But I am a bit confused about the training pipeline of the flow model. **As suggested by [#281 @aluminumbox](https://github.com/FunAudioLLM/CosyVoice/issues/281#issuecomment-2292626433), the flow can be trained just change the param `model` in the run.sh script to 'flow', which will initalize the whole flow model**, e.g., `flow: !new:cosyvoice.flow.flow.MaskedDiffWithXvec` in the cosyvoice.yaml. **I am confused that the loss function of the flow only take the loss of `cosyvoice.flow.flow.MaskedDiffWithXvec.decoder ` into consideration** without any from other modules of the flow (e.g., encoder, length_regulator).
> 
> **_Question:_**
> 
> 1. Is this manner suitable for the flow training considering that the loss is mostly related to the decoder (`cosyvoice.flow.flow.MaskedDiffWithXvec.decoder ` ) of the flow model?  Should we freeze partial of the flow (e.g., the encoder, length_regulator, and nn.Embedding()) to train the flow model?
> 2. How did parameters of other modules (except the decoder) of the flow update if we initialize the whole flow model?
> 
> PS: the reason why I ask this question is that, I found the training of the whole flow model was unstable when initializing the whole flow model. But the training is stable when I only initialized the decoder but freezed other modules by
> 
> ```
> model = configs['flow']
> for param in model.parameters():
>     param.requires_grad = False
> for param in model.decoder.parameters():
>     param.requires_grad = True
> ...
> optimizer = optim.Adam(model.module.decoder.parameters(), **configs['train_conf']['optim_conf'])
> ```
> 
> Here are the loss curves when without (The first one) and with (the second one) freezing. From the first one, it can be observed that there is a crash when the step is around 15k (the generated audio is also worse when steps>15k. But the second one is more stable with the same setting except the freezed module. ![image](https://private-user-images.githubusercontent.com/35420734/360329264-640c0935-394a-4b8c-970f-c150c1bc6ed7.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjg1MjM2MDQsIm5iZiI6MTcyODUyMzMwNCwicGF0aCI6Ii8zNTQyMDczNC8zNjAzMjkyNjQtNjQwYzA5MzUtMzk0YS00YjhjLTk3MGYtYzE1MGMxYmM2ZWQ3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMTAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDEwVDAxMjE0NFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWJlM2RkOGQxYzRjZWYxNTMxZGNlMzg3NTYwZTIxNTcwNTQyNGYwZjgxZmZkNmM0MTFlM2UyYWZkNTg0ZjE3YWQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.MSpWwOcma-YSVIsCJnddd0rz1hgBaKeNxQRsI7ImFMQ) ![image](https://private-user-images.githubusercontent.com/35420734/360329663-cde411c9-60cb-4cd8-98a5-175f657876d3.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjg1MjM2MDQsIm5iZiI6MTcyODUyMzMwNCwicGF0aCI6Ii8zNTQyMDczNC8zNjAzMjk2NjMtY2RlNDExYzktNjBjYi00Y2Q4LTk4YTUtMTc1ZjY1Nzg3NmQzLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMTAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDEwVDAxMjE0NFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWYzMjFiZTU0YjE4NjczNmM2NWJhZDMzOWFlZWU1NTMwMWYxODNiNzU2MmIxY2U4MmNmOWU2ZWQ0ZTIwOWQ4MDkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.UczselZ8WslCGNSgeTgutWv6XEwsz8uLsPlFz6bTy6U)

Hi, huskyachao, what is your progress now? have you got good result to train the flow from scratch? I met the same problem like you, the loss jump back during the warmingup steps.  What is your final lr and batch size of success training?

**Comment by aluminumbox** [comment]
> 关于flow模型的 length_regulator, 没有太明白一些细节。 理解上是由于speech_token抽取的时候，S3以50的码率编码，一个码代表了320个样点； mel谱是用256的步长编码，一个梅尔帧256个样点，16K数据的帧率就是62.5； 现在flow就是要将50码率的东西，转为62.5码率的东西； length_regulator 就是对50的序列内部插值，插值到62.5的码率下长度。感觉length_regulator 里一堆卷积变换的，就是对硬插值后得到的结果做一下平滑。但是length_regulator的inference里，又是分频带又是组合的，这段逻辑在做什么啊？

因为speech token和mel不是整数倍，inference的时候按首中尾分开平滑，是为了减轻流式情况下的overlap处不连续的问题

**Comment by JohnHerry** [comment]
然而好像目前flow的流式合成效果还是不好，还是衔接处有爆音【我是用GT speech token测试的，不是length regulated 的LLM生成token】。 而且我们以前测试过，这个问题不仅仅在flow这边，以前HiFiGAN做流式谱合成时就有衔接处爆音的问题。请问这里hift也是做了什么流式设计的优化吗？

**Comment by echo-hmwang** [comment]
> 正常的情况下 flow模型训练一步需要多少时间啊？需要训练多少步？ 我们这里训练怎么这么慢呢？

hi，请问你用了多少数据训练呢？最后loss收敛到什么范围呢？我目前使用了大概5w小时数据，loss收敛比你放出来的图要慢很多，而且数值也要大

**Comment by JohnHerry** [comment]
建议用10万小时以上数据量训练，请注意数据质量。 训练迭代2个epoch基本可用，如果算力充足建议训练到4个以上。以前测试110W steps效果还是不错的。注意人家开源配置的说明部分，1e-5的学习率是人家用来fine-tune使用的，你从头训练肯定不用这个学习率配置，觉得损失收敛慢就要增加学习率了。另外多卡训练比单卡训练更稳定，损失曲线更漂亮。其它自己多摸索吧。后边总是会踩坑的，坑坑都不同。

**Comment by WelkinYang** [comment]
> > 正常的情况下流程模型训练一步需要多少时间啊？需要训练多少步？我们这里训练怎么这么慢呢？
> 
> 你好，请问你用了多少数据训练呢？最后损失收敛到什么范围呢？我目前用了大概5w小时的数据，损失收敛比你放出来的图要慢很多，而且数值也大

大用大约30万小时的数据从头训练无法恢复现cosy2的说话人相似度自己训练的低很多暂时不知道原因

**Comment by echo-hmwang** [comment]
> > > 正常的情况下流程模型训练一步需要多少时间啊？需要训练多少步？我们这里训练怎么这么慢呢？
> > 
> > 
> > 你好，请问你用了多少数据训练呢？最后损失收敛到什么范围呢？我目前用了大概5w小时的数据，损失收敛比你放出来的图要慢很多，而且数值也大
> 
> 大用大约30万小时的数据从头训练无法恢复现cosy2的说话人相似度自己训练的低很多暂时不知道原因

我改成DiT结构之后重新训了一下，WER和说话人相似度比原版模型稍微好一点，但是流式效果依然不太行。我用的都是公开数据，不确定是不是数据质量的问题

**Comment by WelkinYang** [comment]
> > > > 正常的情况下流程模型训练一步需要多少时间啊？需要训练多少步？我们这里训练怎么这么慢呢？
> > > 
> > > 
> > > 你好，请问你用了多少数据训练呢？最后损失收敛到什么范围呢？我目前用了大概5w小时的数据，损失收敛比你放出来的图要慢很多，而且数值也大
> > 
> > 
> > 大用大约30万小时的数据从头训练无法恢复现cosy2的说话人相似度自己训练的低很多暂时不知道原因
> 
> 我改成DiT结构之后重新训了一下，WER和说话人相似度比原版模型稍微好一点，但是流式效果依然不太行。我用的都是公开数据，不确定是不是数据质量的问题

请问你在seedtts zh eval上的speaker sim大概多少呢  方便交流一下么 我用了30wh的开源数据和1.1b的dit 感觉还是不太对

**Comment by echo-hmwang** [comment]
> > > > > 正常的情况下流程模型训练一步需要多少时间啊？需要训练多少步？我们这里训练怎么这么慢呢？
> > > > 
> > > > 
> > > > 你好，请问你用了多少数据训练呢？最后损失收敛到什么范围呢？我目前用了大概5w小时的数据，损失收敛比你放出来的图要慢很多，而且数值也大
> > > 
> > > 
> > > 大用大约30万小时的数据从头训练无法恢复现cosy2的说话人相似度自己训练的低很多暂时不知道原因
> > 
> > 
> > 我改成DiT结构之后重新训了一下，WER和说话人相似度比原版模型稍微好一点，但是流式效果依然不太行。我用的都是公开数据，不确定是不是数据质量的问题
> 
> 请问你在seedtts zh eval上的speaker sim大概多少呢 方便交流一下么 我用了30wh的开源数据和1.1b的dit 感觉还是不太对

我大概用了5万小时数据和440M的参数量，使用的是F5-TTS的codebase改的，用了22层transformer， 修改的时候保留了cy2里面原有的conformer encoder的上采样部分（ vocoder我改成了bigvgan-v2 24khz版本，所以上采样比例为3.75以匹配vocoder的mel设置）。没有在seedtts zh eval上测。我训练的时候碰到的最大的问题是模型对CFG推理权重很敏感，高频部分容易出现滋滋声，然后改成了和step audio editx类似的结构，在attention后面加了cnn层做平滑部分缓解了这个问题。我记得之前这个repo里面有一个issue提到他们使用了大概16w小时的内部数据训练flow这一部分，我自己训练的版本略有提升，但是参数量相较于cy2翻了好几倍，但是提升效果很小。

**Comment by JohnHerry** [comment]
请问哪位有 flow3 模型的tensorrt 转换脚本？ 我们参照CosyVoice2的tensorrt转换的fp16模型，推理时总是数值溢出。

**Comment by JohnHerry** [comment]
cosy3 flow 的默认配置，训练出来音色相似度还不如cosy2的 flow；还不如自己做的。

**Comment by CJY1018** [comment]
> > > > > > 正常的情况下流程模型训练一步需要多少时间啊？需要训练多少步？我们这里训练怎么这么慢呢？
> > > > > 
> > > > > 
> > > > > 你好，请问你用了多少数据训练呢？最后损失收敛到什么范围呢？我目前用了大概5w小时的数据，损失收敛比你放出来的图要慢很多，而且数值也大
> > > > 
> > > > 
> > > > 大用大约30万小时的数据从头训练无法恢复现cosy2的说话人相似度自己训练的低很多暂时不知道原因
> > > 
> > > 
> > > 我改成DiT结构之后重新训了一下，WER和说话人相似度比原版模型稍微好一点，但是流式效果依然不太行。我用的都是公开数据，不确定是不是数据质量的问题
> > 
> > 
> > 请问你在seedtts zh eval上的speaker sim大概多少呢 方便交流一下么 我用了30wh的开源数据和1.1b的dit 感觉还是不太对
> 
> 我大概用了5万小时数据和440M的参数量，使用的是F5-TTS的codebase改的，用了22层transformer， 修改的时候保留了cy2里面原有的conformer encoder的上采样部分（ vocoder我改成了bigvgan-v2 24khz版本，所以上采样比例为3.75以匹配vocoder的mel设置）。没有在seedtts zh eval上测。我训练的时候碰到的最大的问题是模型对CFG推理权重很敏感，高频部分容易出现滋滋声，然后改成了和step audio editx类似的结构，在attention后面加了cnn层做平滑部分缓解了这个问题。我记得之前这个repo里面有一个issue提到他们使用了大概16w小时的内部数据训练flow这一部分，我自己训练的版本略有提升，但是参数量相较于cy2翻了好几倍，但是提升效果很小。

@echo-hmwang 请问您用的是 [bigvgan_v2_24khz_100band_256x](https://huggingface.co/nvidia/bigvgan_v2_24khz_100band_256x) 这个模型吗？我我发现CosyVoice 2的HiFt Vocoder是 24kHz / 80 mel bands /  480 upsampling ratio 的，而这个 BigVGAN_v2 是 24kHz / 100 mel bands /  256 upsampling ratio 的，请问您是怎么做适配的？
